File: internal/scaffold/cmake.go
package scaffold

import (
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"

	"cup/internal/ui"
)

// ReadFileLines exposes a file's lines to callers outside this package (e.g. the
// dependency scanner). The bool is false only if the file could not be read.
func ReadFileLines(path string) ([]string, bool) {
	return readLines(path)
}

// RemoveDir deletes a directory tree, ignoring a missing path.
func RemoveDir(path string) {
	_ = os.RemoveAll(path)
}

func readLines(path string) ([]string, bool) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, false
	}
	text := string(b)
	if text == "" {
		return nil, true
	}
	return strings.Split(strings.TrimRight(text, "\n"), "\n"), true
}

func writeLines(path string, lines []string) error {
	if len(lines) == 0 {
		return os.WriteFile(path, []byte(""), 0o644)
	}
	return os.WriteFile(path, []byte(strings.Join(lines, "\n")+"\n"), 0o644)
}

// EnsureLine appends line to a CMakeLists (creating it if absent) unless already
// present, verbatim.
func EnsureLine(root, path, line string) error {
	lines, _ := readLines(path)
	for _, l := range lines {
		if l == line {
			return nil
		}
	}
	lines = append(lines, line)
	if err := writeLines(path, lines); err != nil {
		return err
	}
	ui.Updated(fmt.Sprintf("%s  (+ %s)", rel(root, path), line))
	return nil
}

// EnsureLineBefore inserts line immediately before the first line equal to
// anchor, keeping ordering-sensitive directives correct (e.g. third_party must
// precede src/libs). Falls back to appending if the anchor is absent.
func EnsureLineBefore(root, path, line, anchor string) error {
	lines, _ := readLines(path)
	for _, l := range lines {
		if l == line {
			return nil
		}
	}
	inserted := false
	out := make([]string, 0, len(lines)+1)
	for _, l := range lines {
		if !inserted && strings.TrimSpace(l) == anchor {
			out = append(out, line)
			inserted = true
		}
		out = append(out, l)
	}
	if !inserted {
		out = append(out, line)
	}
	if err := writeLines(path, out); err != nil {
		return err
	}
	ui.Updated(fmt.Sprintf("%s  (+ %s)", rel(root, path), line))
	return nil
}

// RemoveLine deletes every occurrence of the exact line. Returns true if any was
// removed.
func RemoveLine(root, path, line string) (bool, error) {
	lines, ok := readLines(path)
	if !ok {
		return false, nil
	}
	kept := lines[:0:0]
	for _, l := range lines {
		if l != line {
			kept = append(kept, l)
		}
	}
	if len(kept) == len(lines) {
		return false, nil
	}
	if err := writeLines(path, kept); err != nil {
		return false, err
	}
	ui.Removed(fmt.Sprintf("%s  (- %s)", rel(root, path), line))
	return true, nil
}

// RemoveMatchingLine deletes every line whose trimmed text matches pattern at its
// start. Used where the registered line may carry extra arguments.
func RemoveMatchingLine(root, path string, pattern *regexp.Regexp) (bool, error) {
	lines, ok := readLines(path)
	if !ok {
		return false, nil
	}
	kept := lines[:0:0]
	for _, l := range lines {
		if !pattern.MatchString(strings.TrimSpace(l)) {
			kept = append(kept, l)
		}
	}
	if len(kept) == len(lines) {
		return false, nil
	}
	if err := writeLines(path, kept); err != nil {
		return false, err
	}
	ui.Removed(rel(root, path))
	return true, nil
}

// AppendBlock appends a multi-line block unless marker already appears in the
// file, treating a repeat registration as a no-op.
func AppendBlock(root, path, marker, block string) error {
	b, _ := os.ReadFile(path)
	existing := string(b)
	if strings.Contains(existing, marker) {
		ui.Skipped(fmt.Sprintf("%s already declares %s", rel(root, path), marker))
		return nil
	}
	prefix := existing
	if prefix != "" && !strings.HasSuffix(prefix, "\n") {
		prefix += "\n"
	}
	if prefix != "" && !strings.HasSuffix(prefix, "\n\n") {
		prefix += "\n"
	}
	if err := os.WriteFile(path, []byte(prefix+block), 0o644); err != nil {
		return err
	}
	ui.Updated(fmt.Sprintf("%s  (+ %s)", rel(root, path), marker))
	return nil
}

// RemoveFetchContentBlock removes the FetchContent_Declare(name …) …
// FetchContent_MakeAvailable(name) block, the inverse of a cmake-download
// registration. Returns true if a block was removed.
func RemoveFetchContentBlock(root, path, name string) (bool, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return false, nil
	}
	pattern := regexp.MustCompile(
		`\n*FetchContent_Declare\(\s*\n\s*` + regexp.QuoteMeta(name) +
			`\b(?s:.*?)FetchContent_MakeAvailable\(\s*` + regexp.QuoteMeta(name) + `\s*\)[ \t]*\n?`,
	)
	newContent, n := replaceCount(pattern, string(b), "\n")
	if n == 0 {
		return false, nil
	}
	if err := os.WriteFile(path, []byte(newContent), 0o644); err != nil {
		return false, err
	}
	ui.Removed(fmt.Sprintf("%s  (- %s)", rel(root, path), name))
	return true, nil
}

func replaceCount(re *regexp.Regexp, src, repl string) (string, int) {
	n := len(re.FindAllString(src, -1))
	if n == 0 {
		return src, 0
	}
	return re.ReplaceAllString(src, repl), n
}

// AddModuleSource appends filename to a library's FILE_SET CXX_MODULES FILES
// block, matching the indentation of the first entry. Idempotent.
func AddModuleSource(root, path, filename string) error {
	b, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("cannot update %s: not found", rel(root, path))
	}
	content := string(b)
	pattern := regexp.MustCompile(`(FILE_SET\s+CXX_MODULES\s+FILES\n)((?:[ \t]+[^\n]+\n)+)`)
	loc := pattern.FindStringSubmatchIndex(content)
	if loc == nil {
		return fmt.Errorf("cannot find FILE_SET CXX_MODULES block in %s", rel(root, path))
	}
	filesBlock := content[loc[4]:loc[5]]
	for _, l := range strings.Split(filesBlock, "\n") {
		if strings.TrimSpace(l) == filename {
			return nil // already listed
		}
	}
	first := strings.SplitN(filesBlock, "\n", 2)[0]
	indent := first[:len(first)-len(strings.TrimLeft(first, " \t"))]
	newBlock := filesBlock + indent + filename + "\n"
	updated := content[:loc[4]] + newBlock + content[loc[5]:]
	if err := os.WriteFile(path, []byte(updated), 0o644); err != nil {
		return err
	}
	ui.Updated(fmt.Sprintf("%s  (+ %s)", rel(root, path), filename))
	return nil
}

// AddHeaderSource appends filename to a header library's FILE_SET HEADERS FILES
// block, matching the indentation of the first entry. Idempotent. It is the
// header-family analogue of AddModuleSource.
func AddHeaderSource(root, path, filename string) error {
	b, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("cannot update %s: not found", rel(root, path))
	}
	content := string(b)
	pattern := regexp.MustCompile(`(FILE_SET\s+HEADERS(?s:.*?)FILES\n)((?:[ \t]+[^\n]+\n)+)`)
	loc := pattern.FindStringSubmatchIndex(content)
	if loc == nil {
		return fmt.Errorf("cannot find FILE_SET HEADERS block in %s", rel(root, path))
	}
	filesBlock := content[loc[4]:loc[5]]
	for _, l := range strings.Split(filesBlock, "\n") {
		if strings.TrimSpace(l) == filename {
			return nil // already listed
		}
	}
	first := strings.SplitN(filesBlock, "\n", 2)[0]
	indent := first[:len(first)-len(strings.TrimLeft(first, " \t"))]
	newBlock := filesBlock + indent + filename + "\n"
	updated := content[:loc[4]] + newBlock + content[loc[5]:]
	if err := os.WriteFile(path, []byte(updated), 0o644); err != nil {
		return err
	}
	ui.Updated(fmt.Sprintf("%s  (+ %s)", rel(root, path), filename))
	return nil
}

var interfaceKeyword = regexp.MustCompile(`\bINTERFACE\b`)

// EnsureHeaderLibStatic promotes a header library's CMakeLists from an INTERFACE
// library to a STATIC one so it can compile a .cpp source. A header-only lib is
// declared `add_library(<name> INTERFACE)` with INTERFACE-scoped properties;
// adding the first compiled component turns it into `add_library(<name> STATIC)`
// with those properties rescoped to PUBLIC. A no-op if the lib is already STATIC.
func EnsureHeaderLibStatic(root, path, name string) error {
	b, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("cannot update %s: not found", rel(root, path))
	}
	content := string(b)
	interfaceDecl := fmt.Sprintf("add_library(%s INTERFACE)", name)
	if !strings.Contains(content, interfaceDecl) {
		return nil // already STATIC (or not a header lib we manage)
	}
	content = strings.Replace(content, interfaceDecl, fmt.Sprintf("add_library(%s STATIC)", name), 1)
	content = interfaceKeyword.ReplaceAllString(content, "PUBLIC")
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return err
	}
	ui.Updated(fmt.Sprintf("%s  (INTERFACE -> STATIC)", rel(root, path)))
	return nil
}

// AddPartitionImport re-exports a module partition from its lib's primary module
// interface unit, inserting `export import :<partition>;` — appended to the
// existing block of partition imports, or as a fresh block right after the
// `export module` declaration. Idempotent.
func AddPartitionImport(root, primary, partition string) error {
	b, err := os.ReadFile(primary)
	if err != nil {
		return fmt.Errorf("cannot update %s: not found", rel(root, primary))
	}
	directive := fmt.Sprintf("export import :%s;", partition)
	lines := strings.Split(strings.TrimRight(string(b), "\n"), "\n")
	for _, l := range lines {
		if l == directive {
			return nil
		}
	}
	out, err := insertPartitionImport(lines, directive)
	if err != nil {
		return fmt.Errorf("%w in %s", err, rel(root, primary))
	}
	if err := os.WriteFile(primary, []byte(strings.Join(out, "\n")+"\n"), 0o644); err != nil {
		return err
	}
	ui.Updated(fmt.Sprintf("%s  (+ %s)", rel(root, primary), directive))
	return nil
}

// insertPartitionImport returns lines with directive inserted after the last
// existing `export import :` line, or — if there are none — as a fresh block
// right after the `export module` declaration. It errors if no module
// declaration is present.
func insertPartitionImport(lines []string, directive string) ([]string, error) {
	lastImport := -1
	for i, l := range lines {
		if strings.HasPrefix(l, "export import :") {
			lastImport = i
		}
	}
	if lastImport >= 0 {
		out := append([]string{}, lines[:lastImport+1]...)
		out = append(out, directive)
		return append(out, lines[lastImport+1:]...), nil
	}
	for i, l := range lines {
		if strings.HasPrefix(l, "export module ") {
			out := append([]string{}, lines[:i+1]...)
			out = append(out, "", directive)
			return append(out, lines[i+1:]...), nil
		}
	}
	return nil, fmt.Errorf("cannot find module declaration")
}

// ListSubdirs returns the sorted names of immediate subdirectories of path.
func ListSubdirs(path string) []string {
	entries, err := os.ReadDir(path)
	if err != nil {
		return nil
	}
	var dirs []string
	for _, e := range entries {
		if e.IsDir() {
			dirs = append(dirs, e.Name())
		}
	}
	sort.Strings(dirs)
	return dirs
}

File: cpp/src/libs/cup/scaffold/Cmake.cpp
module;
#include <algorithm>
#include <expected>
#include <filesystem>
#include <format>
#include <fstream>
#include <ios>
#include <iterator>
#include <optional>
#include <regex>
#include <string>
#include <string_view>
#include <system_error>
#include <utility>
#include <vector>
module cup.scaffold;

import cup.ui;

namespace cup::scaffold {
namespace {

// split mirrors Go's strings.Split(s, sep) for a single-byte separator,
// including its two easy-to-miss edge cases: an empty s yields one empty
// element (not zero), and a trailing sep yields a trailing empty element.
std::vector<std::string> split(std::string_view s, char sep) {
    std::vector<std::string> parts;
    std::size_t start = 0;
    while (true) {
        const auto pos = s.find(sep, start);
        if (pos == std::string_view::npos) {
            parts.emplace_back(s.substr(start));
            break;
        }
        parts.emplace_back(s.substr(start, pos - start));
        start = pos + 1;
    }
    return parts;
}

std::string trim(std::string_view s) {
    constexpr std::string_view kSpace = " \t\n\r\f\v";
    const auto first = s.find_first_not_of(kSpace);
    if (first == std::string_view::npos) {
        return {};
    }
    const auto last = s.find_last_not_of(kSpace);
    return std::string(s.substr(first, last - first + 1));
}

std::optional<std::string> slurp(const std::filesystem::path& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) {
        return std::nullopt;
    }
    return std::string(std::istreambuf_iterator<char>(in), std::istreambuf_iterator<char>());
}

std::expected<void, error::Error> write_bytes(const std::filesystem::path& path,
                                              std::string_view content) {
    std::ofstream out(path, std::ios::binary | std::ios::trunc);
    out << content;
    if (!out) {
        return std::unexpected(error::Error(std::format("writing {}", path.string())));
    }
    return {};
}

std::optional<std::vector<std::string>> read_lines(const std::filesystem::path& path) {
    auto text = slurp(path);
    if (!text.has_value()) {
        return std::nullopt;
    }
    if (text->empty()) {
        return std::vector<std::string>{};
    }
    while (!text->empty() && text->back() == '\n') {
        text->pop_back();
    }
    return split(*text, '\n');
}

std::expected<void, error::Error> write_lines(const std::filesystem::path& path,
                                              const std::vector<std::string>& lines) {
    if (lines.empty()) {
        return write_bytes(path, "");
    }
    std::string content;
    for (std::size_t i = 0; i < lines.size(); ++i) {
        if (i != 0) {
            content += '\n';
        }
        content += lines[i];
    }
    content += '\n';
    return write_bytes(path, content);
}

std::string quote_meta(std::string_view s) {
    constexpr std::string_view kSpecial = R"(\^$.|?*+()[]{})";
    std::string out;
    out.reserve(s.size());
    for (const char c : s) {
        if (kSpecial.contains(c)) {
            out += '\\';
        }
        out += c;
    }
    return out;
}

std::pair<std::string, int> replace_count(const std::regex& re, const std::string& src,
                                          const std::string& repl) {
    const auto n = static_cast<int>(
        std::distance(std::sregex_iterator(src.begin(), src.end(), re), std::sregex_iterator()));
    if (n == 0) {
        return {src, 0};
    }
    return {std::regex_replace(src, re, repl), n};
}

// Shared by add_module_source / add_header_source: both look for a two-group
// pattern (the FILE_SET header, then the indented block of file entries),
// insert filename into the second group at the first entry's indentation, and
// leave everything else untouched.
std::expected<void, error::Error> add_file_set_entry(const std::filesystem::path& root, const std::filesystem::path& path,
                                                      std::string_view filename,
                                                      const std::regex& pattern,
                                                      std::string_view block_label) {
    auto content = slurp(path);
    if (!content.has_value()) {
        return std::unexpected(error::Error(std::format("cannot update {}: not found", rel(root, path))));
    }
    std::smatch match;
    if (!std::regex_search(*content, match, pattern)) {
        return std::unexpected(
            error::Error(std::format("cannot find {} block in {}", block_label, rel(root, path))));
    }
    const std::string files_block = match[2].str();
    for (const auto& l : split(files_block, '\n')) {
        if (trim(l) == filename) {
            return {}; // already listed
        }
    }
    const std::string first = files_block.substr(0, files_block.find('\n'));
    const auto indent_end = first.find_first_not_of(" \t");
    const std::string indent = first.substr(0, indent_end == std::string::npos ? first.size() : indent_end);
    const std::string new_block = files_block + indent + std::string(filename) + "\n";
    const auto group2_start = static_cast<std::size_t>(match.position(2));
    const auto group2_end = group2_start + static_cast<std::size_t>(match.length(2));
    const std::string updated = content->substr(0, group2_start) + new_block + content->substr(group2_end);
    if (auto written = write_bytes(path, updated); !written.has_value()) {
        return written;
    }
    ui::updated(std::format("{}  (+ {})", rel(root, path), filename));
    return {};
}

std::optional<std::vector<std::string>> insert_partition_import(std::vector<std::string> lines,
                                                                 std::string_view directive) {
    int last_import = -1;
    for (std::size_t i = 0; i < lines.size(); ++i) {
        if (lines[i].starts_with("export import :")) {
            last_import = static_cast<int>(i);
        }
    }
    if (last_import >= 0) {
        lines.insert(lines.begin() + last_import + 1, std::string(directive));
        return lines;
    }
    if (const auto module_decl = std::ranges::find_if(
            lines, [](const std::string& l) { return l.starts_with("export module "); });
        module_decl != lines.end()) {
        lines.insert(module_decl + 1, {std::string(), std::string(directive)});
        return lines;
    }
    return std::nullopt;
}

std::string join_lines(const std::vector<std::string>& lines) {
    std::string out;
    for (std::size_t i = 0; i < lines.size(); ++i) {
        if (i != 0) {
            out += '\n';
        }
        out += lines[i];
    }
    out += '\n';
    return out;
}

}

std::optional<std::vector<std::string>> read_file_lines(const std::filesystem::path& path) {
    return read_lines(path);
}

void remove_dir(const std::filesystem::path& path) {
    std::error_code ec;
    std::filesystem::remove_all(path, ec);
}

std::expected<void, error::Error> ensure_line(const std::filesystem::path& root, const std::filesystem::path& path,
                                              std::string_view line) {
    auto lines = read_lines(path).value_or(std::vector<std::string>{});
    for (const auto& l : lines) {
        if (l == line) {
            return {};
        }
    }
    lines.emplace_back(line);
    if (auto written = write_lines(path, lines); !written.has_value()) {
        return written;
    }
    ui::updated(std::format("{}  (+ {})", rel(root, path), line));
    return {};
}

std::expected<void, error::Error> ensure_line_before(const std::filesystem::path& root, const std::filesystem::path& path,
                                                      std::string_view line, std::string_view anchor) {
    auto lines = read_lines(path).value_or(std::vector<std::string>{});
    for (const auto& l : lines) {
        if (l == line) {
            return {};
        }
    }
    std::vector<std::string> out;
    out.reserve(lines.size() + 1);
    bool inserted = false;
    for (const auto& l : lines) {
        if (!inserted && trim(l) == anchor) {
            out.emplace_back(line);
            inserted = true;
        }
        out.push_back(l);
    }
    if (!inserted) {
        out.emplace_back(line);
    }
    if (auto written = write_lines(path, out); !written.has_value()) {
        return written;
    }
    ui::updated(std::format("{}  (+ {})", rel(root, path), line));
    return {};
}

std::expected<bool, error::Error> remove_line(const std::filesystem::path& root, const std::filesystem::path& path,
                                              std::string_view line) {
    auto maybe_lines = read_lines(path);
    if (!maybe_lines.has_value()) {
        return false;
    }
    std::vector<std::string> kept;
    kept.reserve(maybe_lines->size());
    for (const auto& l : *maybe_lines) {
        if (l != line) {
            kept.push_back(l);
        }
    }
    if (kept.size() == maybe_lines->size()) {
        return false;
    }
    if (auto written = write_lines(path, kept); !written.has_value()) {
        return std::unexpected(std::move(written).error());
    }
    ui::removed(std::format("{}  (- {})", rel(root, path), line));
    return true;
}

std::expected<bool, error::Error> remove_matching_line(const std::filesystem::path& root, const std::filesystem::path& path,
                                                        std::string_view pattern) {
    auto maybe_lines = read_lines(path);
    if (!maybe_lines.has_value()) {
        return false;
    }
    const std::regex re{std::string(pattern)};
    std::vector<std::string> kept;
    kept.reserve(maybe_lines->size());
    for (const auto& l : *maybe_lines) {
        if (!std::regex_search(trim(l), re)) {
            kept.push_back(l);
        }
    }
    if (kept.size() == maybe_lines->size()) {
        return false;
    }
    if (auto written = write_lines(path, kept); !written.has_value()) {
        return std::unexpected(std::move(written).error());
    }
    ui::removed(rel(root, path));
    return true;
}

std::expected<void, error::Error> append_block(const std::filesystem::path& root, const std::filesystem::path& path,
                                               std::string_view marker, std::string_view block) {
    const std::string existing = slurp(path).value_or(std::string());
    if (existing.contains(marker)) {
        ui::skipped(std::format("{} already declares {}", rel(root, path), marker));
        return {};
    }
    std::string prefix = existing;
    if (!prefix.empty() && !prefix.ends_with('\n')) {
        prefix += '\n';
    }
    if (!prefix.empty() && !prefix.ends_with("\n\n")) {
        prefix += '\n';
    }
    if (auto written = write_bytes(path, prefix + std::string(block)); !written.has_value()) {
        return written;
    }
    ui::updated(std::format("{}  (+ {})", rel(root, path), marker));
    return {};
}

std::expected<bool, error::Error> remove_fetch_content_block(const std::filesystem::path& root,
                                                              const std::filesystem::path& path,
                                                              std::string_view name) {
    auto content = slurp(path);
    if (!content.has_value()) {
        return false;
    }
    const std::string pattern = "\\n*FetchContent_Declare\\(\\s*\\n\\s*" + quote_meta(name) +
                                "\\b[\\s\\S]*?FetchContent_MakeAvailable\\(\\s*" + quote_meta(name) +
                                "\\s*\\)[ \\t]*\\n?";
    const std::regex re(pattern);
    auto [new_content, n] = replace_count(re, *content, "\n");
    if (n == 0) {
        return false;
    }
    if (auto written = write_bytes(path, new_content); !written.has_value()) {
        return std::unexpected(std::move(written).error());
    }
    ui::removed(std::format("{}  (- {})", rel(root, path), name));
    return true;
}

std::expected<void, error::Error> add_module_source(const std::filesystem::path& root, const std::filesystem::path& path,
                                                     std::string_view filename) {
    static const std::regex pattern(R"((FILE_SET\s+CXX_MODULES\s+FILES\n)((?:[ \t]+[^\n]+\n)+))");
    return add_file_set_entry(root, path, filename, pattern, "FILE_SET CXX_MODULES");
}

std::expected<void, error::Error> add_header_source(const std::filesystem::path& root, const std::filesystem::path& path,
                                                     std::string_view filename) {
    static const std::regex pattern(R"((FILE_SET\s+HEADERS[\s\S]*?FILES\n)((?:[ \t]+[^\n]+\n)+))");
    return add_file_set_entry(root, path, filename, pattern, "FILE_SET HEADERS");
}

std::expected<void, error::Error> ensure_header_lib_static(const std::filesystem::path& root, const std::filesystem::path& path,
                                                            std::string_view name) {
    auto content = slurp(path);
    if (!content.has_value()) {
        return std::unexpected(error::Error(std::format("cannot update {}: not found", rel(root, path))));
    }
    const std::string interface_decl = std::format("add_library({} INTERFACE)", name);
    const auto pos = content->find(interface_decl);
    if (pos == std::string::npos) {
        return {}; // already STATIC (or not a header lib we manage)
    }
    content->replace(pos, interface_decl.size(), std::format("add_library({} STATIC)", name));
    static const std::regex interface_keyword(R"(\bINTERFACE\b)");
    const std::string updated = std::regex_replace(*content, interface_keyword, "PUBLIC");
    if (auto written = write_bytes(path, updated); !written.has_value()) {
        return written;
    }
    ui::updated(std::format("{}  (INTERFACE -> STATIC)", rel(root, path)));
    return {};
}

std::expected<void, error::Error> add_partition_import(const std::filesystem::path& root, const std::filesystem::path& primary,
                                                        std::string_view partition) {
    auto text = slurp(primary);
    if (!text.has_value()) {
        return std::unexpected(error::Error(std::format("cannot update {}: not found", rel(root, primary))));
    }
    while (!text->empty() && text->back() == '\n') {
        text->pop_back();
    }
    const std::string directive = std::format("export import :{};", partition);
    auto lines = split(*text, '\n');
    for (const auto& l : lines) {
        if (l == directive) {
            return {};
        }
    }
    auto out = insert_partition_import(std::move(lines), directive);
    if (!out.has_value()) {
        return std::unexpected(
            error::Error(std::format("cannot find module declaration in {}", rel(root, primary))));
    }
    if (auto written = write_bytes(primary, join_lines(*out)); !written.has_value()) {
        return written;
    }
    ui::updated(std::format("{}  (+ {})", rel(root, primary), directive));
    return {};
}

std::vector<std::string> list_subdirs(const std::filesystem::path& path) {
    std::error_code ec;
    if (!std::filesystem::is_directory(path, ec)) {
        return {};
    }
    std::vector<std::string> dirs;
    for (const auto& entry : std::filesystem::directory_iterator(path, ec)) {
        if (entry.is_directory()) {
            dirs.push_back(entry.path().filename().string());
        }
    }
    std::ranges::sort(dirs);
    return dirs;
}

}
